include/pybind11/pytypes.h

```diff
@@ -1761,7 +1774,10 @@ class list : public object {
     size_t size() const { return (size_t) PyList_Size(m_ptr); }
     bool empty() const { return size() == 0; }
     detail::list_accessor operator[](size_t index) const { return {*this, index}; }
-    detail::item_accessor operator[](handle h) const { return object::operator[](h); }
+    template <typename T, detail::enable_if_t<detail::is_pyobject<T>::value, int> = 0>
+    detail::item_accessor operator[](T o) const {
```

Comment:
@rwgk Okay, this is a bit more verbose now and uses templates, but at least it's more general.